src/coprocessor/util.rs

```diff
@@ -17,33 +17,76 @@ use tipb::schema::ColumnInfo;
 use super::codec::datum::Datum;
 use super::codec::mysql;
 
-/// Get the smallest key which is larger than the key given.
-pub fn prefix_next(key: &[u8]) -> Vec<u8> {
-    let mut nk = key.to_vec();
-    if nk.is_empty() {
-        nk.push(0);
-        return nk;
+/// Convert the key to the smallest key which is larger than the key given.
+pub fn convert_to_prefix_next(key: &mut Vec<u8>) {
+    if key.is_empty() {
+        key.push(0);
+        return;
     }
-    let mut i = nk.len() - 1;
+    let mut i = key.len() - 1;
+
+    // Add 1 to the last byte that is not 255, and set it's following bytes to 0.
     loop {
-        if nk[i] == 255 {
-            nk[i] = 0;
+        if key[i] == 255 {
+            key[i] = 0;
         } else {
-            nk[i] += 1;
-            return nk;
+            key[i] += 1;
+            return;
         }
         if i == 0 {
-            nk = key.to_vec();
-            nk.push(0);
-            return nk;
+            // All bytes are 255. Append a 0 to the key.
+            for byte in key.iter_mut() {
+                *byte = 255;
+            }
+            key.push(0);
+            return;
         }
         i -= 1;
     }
 }
 
+/// Check if `key`'s prefix next equals to `next`
+pub fn is_prefix_next(key: &[u8], next: &[u8]) -> bool {
```

Comment:
This function is a little hard to understand. How about simplify it by simulating (not real perform) a prefix next? In this case we don't need to change the actual bytes, just check whether they meet the given bytes.

Reply:
I've added some comments in the function. PTAL. Does it help you understanding it?